P03 D Panchal

---
- Python using Pandas and NumPy
---
# Ebay car sales - data cleaning & exploring


We have this dataset from classified section of ebay Germany for used car. It has around 50,000 data points. 
The aim of project is to first clean the data set. And the analyse and find insights into car sales. 

Let us review what it contains, identify issues and try to clean. 

In [62]:
import pandas as pd
import numpy as np
autos = pd.read_csv("autos.csv",  encoding='Latin-1')

In [63]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

### General info about data set
There are *20 columns* with varying data types, some of the columns have missing or null values too as we can see those have less than 50,000 entries in info below. 

In [64]:
autos.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

### First few rows

In [65]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Correct Column labels
These are currently camelcase and not *standard*. We need to convert to the usual snakecase. The new column names should be as below. Let us assign these to the table, replacing the original labels.

In [66]:
new_column_name = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type',
       'year_of_registration', 'gearbox', 'power_ps', 'model', 'odometer',
       'month_of_registration', 'fuel_type', 'brand', 'not_repaired_damage',
       'date_created', 'nr_of_pictures', 'postal_code', 'last_seen']

In [67]:
autos.columns = new_column_name

### First rows again
Now that the labels should have been replaced, we should check first few rows of the table again.

In [68]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  year_of_registration  \
0  privat    Angebot  $5,000  control          bus                  2004   
1  privat    Angebot  $8,500  control    limousine                  1997   
2  privat    Angebot  $8,990     test    limousine                  2009   
3  privat    Angebot  $4,350  control   kleinwagen                  2007   
4  privat    Angebot  $1,350     test        kombi                  2003   

     gearbox  power_ps   model   odometer  month_of_registration fuel_type  \
0    manuell       158  andere  150,000km                      3       lpg   
1  automatik       286     7er  150,000km                      6    benzin   
2    manuell       102    golf   70,000km                      7    benzin   
3  automatik        71  fortwo   70,000km                      6    benzin   
4    manuell         0   focus  150,000km                      7    benzin   

        brand not_repaired_damage         date_created  nr_of_pictures  \
0     peugeot                nein  2016-03-26 00:00:00               0   
1         bmw                nein  2016-04-04 00:00:00               0   
2  volkswagen                nein  2016-03-26 00:00:00               0   
3       smart                nein  2016-03-12 00:00:00               0   
4        ford                nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

**Statistical** description of columns we see below.

In [69]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  year_of_registration  gearbox      power_ps  model  \
count         44905          50000.000000    47320  50000.000000  47242   
unique            8                   NaN        2           NaN    245   
top       limousine                   NaN  manuell           NaN   golf   
freq          12859                   NaN    36993           NaN   4024   
mean            NaN           2005.073280      NaN    116.355920    NaN   
std             NaN            105.712813      NaN    209.216627    NaN   
min             NaN           1000.000000      NaN      0.000000    NaN   
25%             NaN           1999.000000      NaN     70.000000    NaN   
50%             NaN           2003.000000      NaN    105.000000    NaN   
75%             NaN           2008.000000      NaN    150.000000    NaN   
max             NaN           9999.000000      NaN  17700.000000    NaN   

         odometer  month_of_registration fuel_type       brand  \
count       50000           50000.000000     45518       50000   
unique         13                    NaN         7          40   
top     150,000km                    NaN    benzin  volkswagen   
freq        32424                    NaN     30107       10687   
mean          NaN               5.723360       NaN         NaN   
std           NaN               3.711984       NaN         NaN   
min           NaN               0.000000       NaN         NaN   
25%           NaN               3.000000       NaN         NaN   
50%           NaN               6.000000       NaN         NaN   
75%           NaN               9.000000       NaN         NaN   
max           NaN              12.000000       NaN         NaN   

       not_repaired_damage         date_created  nr_of_pictures   postal_code  \
count                40171                50000         50000.0  50000.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 35232                 1946             NaN           NaN   
mean                   NaN                  NaN             0.0  50813.627300   
std                    NaN                  NaN             0.0  25779.747957   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30451.000000   
50%                    NaN                  NaN             0.0  49577.000000   
75%                    NaN                  NaN             0.0  71540.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

### Columns content
Taking closer look at the contents of columns, we note from above that
- **One value** Some columns have mostly one value only. These hardly provide much info, like **offer_type** and **seller**. Both these are dominated by one value. These column hence can be removed.

- **Investigate further** Some columns need to be explored, investigated further. For example year of registration has lowest value of 1000. The number of pictures column does not provide proper count. May be it is not stored as number.

- **Number stored as text** Also, odometer and price seem to be stored as text.

Let us begin to convert these text to number.

### Odometer 


In [70]:
autos.loc[:,'odometer']

0        150,000km
1        150,000km
2         70,000km
3         70,000km
4        150,000km
5        150,000km
6        150,000km
7        150,000km
8        150,000km
9        150,000km
10       150,000km
11       150,000km
12        50,000km
13       150,000km
14       150,000km
15        80,000km
16       150,000km
17       150,000km
18       150,000km
19       150,000km
20       150,000km
21       150,000km
22        10,000km
23       150,000km
24        30,000km
25       150,000km
26       125,000km
27       150,000km
28       150,000km
29       150,000km
           ...    
49970     60,000km
49971    150,000km
49972    150,000km
49973    150,000km
49974    150,000km
49975    100,000km
49976    150,000km
49977    150,000km
49978    150,000km
49979     70,000km
49980    125,000km
49981    150,000km
49982     90,000km
49983    150,000km
49984    150,000km
49985    150,000km
49986    125,000km
49987     50,000km
49988    150,000km
49989    150,000km
49990     30,000km
49991    150

as we can see above we need to **remove** `km` and "`,`" before converting text to number.

In [71]:
autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',','').astype(int)


In [72]:
autos['odometer']

0        150000
1        150000
2         70000
3         70000
4        150000
5        150000
6        150000
7        150000
8        150000
9        150000
10       150000
11       150000
12        50000
13       150000
14       150000
15        80000
16       150000
17       150000
18       150000
19       150000
20       150000
21       150000
22        10000
23       150000
24        30000
25       150000
26       125000
27       150000
28       150000
29       150000
          ...  
49970     60000
49971    150000
49972    150000
49973    150000
49974    150000
49975    100000
49976    150000
49977    150000
49978    150000
49979     70000
49980    125000
49981    150000
49982     90000
49983    150000
49984    150000
49985    150000
49986    125000
49987     50000
49988    150000
49989    150000
49990     30000
49991    150000
49992    125000
49993    150000
49994    150000
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer, Length: 

In [73]:
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

### Price
Similarly let us look at the contents of column `price` and understand how it needs to be **cleaned**.
We need to remove the `$` and `,`.

In [74]:
autos['price']

0         $5,000
1         $8,500
2         $8,990
3         $4,350
4         $1,350
5         $7,900
6           $300
7         $1,990
8           $250
9           $590
10          $999
11          $350
12        $5,299
13        $1,350
14        $3,999
15       $18,900
16          $350
17        $5,500
18          $300
19        $4,150
20        $3,500
21       $41,500
22       $25,450
23        $7,999
24       $48,500
25           $90
26          $777
27            $0
28        $5,250
29        $4,999
          ...   
49970    $15,800
49971       $950
49972     $3,300
49973     $6,000
49974         $0
49975     $9,700
49976     $5,900
49977     $5,500
49978       $900
49979    $11,000
49980       $400
49981     $2,000
49982     $1,950
49983       $600
49984         $0
49985     $1,000
49986    $15,900
49987    $21,990
49988     $9,550
49989       $150
49990    $17,500
49991       $500
49992     $4,800
49993     $1,650
49994     $5,000
49995    $24,900
49996     $1,980
49997    $13,2

In [75]:
autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int)


In [76]:
autos['price']

0         5000
1         8500
2         8990
3         4350
4         1350
5         7900
6          300
7         1990
8          250
9          590
10         999
11         350
12        5299
13        1350
14        3999
15       18900
16         350
17        5500
18         300
19        4150
20        3500
21       41500
22       25450
23        7999
24       48500
25          90
26         777
27           0
28        5250
29        4999
         ...  
49970    15800
49971      950
49972     3300
49973     6000
49974        0
49975     9700
49976     5900
49977     5500
49978      900
49979    11000
49980      400
49981     2000
49982     1950
49983      600
49984        0
49985     1000
49986    15900
49987    21990
49988     9550
49989      150
49990    17500
49991      500
49992     4800
49993     1650
49994     5000
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64

In [77]:
autos.rename({'price': 'price_$'}, axis=1, inplace=True)

### Rename Column

The odometer and price columns have now been converted to numbers from text. And also the column label has been renamed as `odometer_km` and `price_$`

### Analyse values inside Odometer

Let us see values and frequency of different values inside Odometer. This will help to find outliers or extreme values.

In [78]:
autos['odometer_km'].describe()


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [79]:
autos['odometer_km'].unique()

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000])

In [80]:
Odometer_freq = autos['odometer_km'].value_counts()

### High to low odometer values
The range and distribution of values seems to be as expected. There are no exceptional values, as seen in below.

In [81]:
# sort values
Odometer_freq.sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

In [82]:
autos['price_$'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64

In [83]:
price_values = autos['price_$'].unique()

In [84]:
price_freq = autos['price_$'].value_counts()
# sort price for review
price_freq.sort_index(ascending=False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
3890000        1
1300000        1
1234566        1
999999         2
999990         1
350000         1
345000         1
299000         1
295000         1
265000         1
259000         1
250000         1
220000         1
198000         1
197000         1
194000         1
190000         1
180000         1
175000         1
169999         1
169000         1
163991         1
163500         1
155000         1
151990         1
            ... 
66             1
65             5
60             9
59             1
55             2
50            49
49             4
47             1
45             4
40             6
35             1
30             7
29             1
25             5
20             4
18             1
17             3
15             2
14             1
13             2
12             3
11             2
10             7
9              1
8              1
5              2
3              1
2             

### Price unreasonable datapoints
With price as `0` there are *1421* entries. These entries dont make sense. 

Similarly, on the higher side there are some values, for example below entries seem to be unreasonable. These seem more like typing errors or carless entries.

    99999999      
    27322222      
    12345678      
    11111111      
    10000000       

Therefore `3890000` should be highest *acceptable value*. All values **above** `3890000` should be **removed**. 

In [85]:
# remove rows with unreasonable prices
new_autos = autos[(autos['price_$'] > 3890000) | (autos['price_$'] != 0)]

### Remaining cleaned rows
After removing the zero price as well as abnormally high price values, from 50,000 entries, we are **left with** `48579` final values. These are saved as `new_autos` for time being. 

In [86]:
new_autos.describe(include='all')

date_crawled         name  seller offer_type       price_$  \
count                 48579        48579   48579      48579  4.857900e+04   
unique                46895        37481       2          1           NaN   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           76   48578      48579           NaN   
mean                    NaN          NaN     NaN        NaN  1.012788e+04   
std                     NaN          NaN     NaN        NaN  4.880873e+05   
min                     NaN          NaN     NaN        NaN  1.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.200000e+03   
50%                     NaN          NaN     NaN        NaN  3.000000e+03   
75%                     NaN          NaN     NaN        NaN  7.490000e+03   
max                     NaN          NaN     NaN        NaN  1.000000e+08   

       abtest vehicle_type  year_of_registration  gearbox      power_ps  \
count   48579        43990          48579.000000    46232  48579.000000   
unique      2            8                   NaN        2           NaN   
top      test    limousine                   NaN  manuell           NaN   
freq    25025        12603                   NaN    36110           NaN   
mean      NaN          NaN           2004.753000      NaN    117.212067   
std       NaN          NaN             88.631663      NaN    200.650589   
min       NaN          NaN           1000.000000      NaN      0.000000   
25%       NaN          NaN           1999.000000      NaN     71.000000   
50%       NaN          NaN           2004.000000      NaN    107.000000   
75%       NaN          NaN           2008.000000      NaN    150.000000   
max       NaN          NaN           9999.000000      NaN  17700.000000   

        model    odometer_km  month_of_registration fuel_type       brand  \
count   46116   48579.000000           48579.000000     44544       48579   
unique    245            NaN                    NaN         7          40   
top      golf            NaN                    NaN    benzin  volkswagen   
freq     3900            NaN                    NaN     29375       10339   
mean      NaN  125766.174685               5.781860       NaN         NaN   
std       NaN   39792.936148               3.685759       NaN         NaN   
min       NaN    5000.000000               0.000000       NaN         NaN   
25%       NaN  125000.000000               3.000000       NaN         NaN   
50%       NaN  150000.000000               6.000000       NaN         NaN   
75%       NaN  150000.000000               9.000000       NaN         NaN   
max       NaN  150000.000000              12.000000       NaN         NaN   

       not_repaired_damage         date_created  nr_of_pictures   postal_code  \
count                39472                48579         48579.0  48579.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 34782                 1887             NaN           NaN   
mean                   NaN                  NaN             0.0  50976.545400   
std                    NaN                  NaN             0.0  25749.386337   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30657.000000   
50%                    NaN                  NaN             0.0  49716.000000   
75%                    NaN                  NaN             0.0  71665.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 48579  
unique                38483  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                  

### Dates columns
We have five columns containing date information. 
- `date_crawled`: added by the crawler 
- `last_seen`: added by the crawler 
- `ad_created`: from the website
- `registration_month`: from the website*
- `registration_year`: from the website*
`*` = these are already stored as number

#### Month and Year
The *month* and *year* columns can be analysed straighaway as they are numerical already. Let us review in below. The rest of columns will have to be formatted for analysis.
 

In [87]:
new_autos['month_of_registration'].describe()

count    48579.000000
mean         5.781860
std          3.685759
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: month_of_registration, dtype: float64

In [88]:
new_autos['month_of_registration'].value_counts(ascending=False)

3     5005
0     4482
6     4272
4     4037
5     4031
7     3858
10    3588
12    3376
9     3331
11    3313
1     3221
8     3126
2     2939
Name: month_of_registration, dtype: int64

### Month
There are 4482, about 10% of total entries, that have **month as `0` zero**. These entries are ones most likely to be not filled by user, or filled as zero as actual value may not be available to fill in.

In [89]:
new_autos['year_of_registration'].describe()

count    48579.000000
mean      2004.753000
std         88.631663
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: year_of_registration, dtype: float64

In [90]:
new_autos['year_of_registration'].value_counts(ascending=True)

1952       1
1931       1
1000       1
1948       1
9000       1
1927       1
8888       1
4100       1
6200       1
2800       1
1929       1
1938       1
1943       1
5911       1
4800       1
1111       1
1953       1
4500       1
1939       1
1001       1
2019       2
1955       2
1957       2
1954       2
1934       2
1941       2
1951       2
1800       2
9999       3
1950       3
        ... 
1989     174
1991     339
1990     347
1992     371
2015     392
1993     425
2018     471
1994     629
2014     664
2013     803
2016    1220
1995    1228
2012    1310
1996    1373
2017    1393
2010    1589
2011    1623
1997    1951
2009    2086
2008    2215
2007    2277
1998    2363
2002    2486
2001    2637
2006    2671
2003    2699
2004    2703
1999    2899
2005    2936
2000    3156
Name: year_of_registration, Length: 95, dtype: int64

In [91]:
# years = new_autos['year_of_registration'].unique()
new_autos.describe()

price_$  year_of_registration      power_ps    odometer_km  \
count  4.857900e+04          48579.000000  48579.000000   48579.000000   
mean   1.012788e+04           2004.753000    117.212067  125766.174685   
std    4.880873e+05             88.631663    200.650589   39792.936148   
min    1.000000e+00           1000.000000      0.000000    5000.000000   
25%    1.200000e+03           1999.000000     71.000000  125000.000000   
50%    3.000000e+03           2004.000000    107.000000  150000.000000   
75%    7.490000e+03           2008.000000    150.000000  150000.000000   
max    1.000000e+08           9999.000000  17700.000000  150000.000000   

       month_of_registration  nr_of_pictures   postal_code  
count           48579.000000         48579.0  48579.000000  
mean                5.781860             0.0  50976.545400  
std                 3.685759             0.0  25749.386337  
min                 0.000000             0.0   1067.000000  
25%                 3.000000             0.0  30657.000000  
50%                 6.000000             0.0  49716.000000  
75%                 9.000000             0.0  71665.000000  
max                12.000000             0.0  99998.000000

In [92]:
# years_b = years[(years >= 1910 )& (years <= 2019)]
bool_for_new2 = (new_autos['year_of_registration'] >= 1910) & (new_autos['year_of_registration'] <= 2019)

new2_autos = new_autos[bool_for_new2]

In [93]:
new2_autos.describe(include='all')

date_crawled         name  seller offer_type       price_$  \
count                 48559        48559   48559      48559  4.855900e+04   
unique                46876        37463       2          1           NaN   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           76   48558      48559           NaN   
mean                    NaN          NaN     NaN        NaN  1.012912e+04   
std                     NaN          NaN     NaN        NaN  4.881877e+05   
min                     NaN          NaN     NaN        NaN  1.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.200000e+03   
50%                     NaN          NaN     NaN        NaN  3.000000e+03   
75%                     NaN          NaN     NaN        NaN  7.490000e+03   
max                     NaN          NaN     NaN        NaN  1.000000e+08   

       abtest vehicle_type  year_of_registration  gearbox      power_ps  \
count   48559        43990          48559.000000    46224  48559.000000   
unique      2            8                   NaN        2           NaN   
top      test    limousine                   NaN  manuell           NaN   
freq    25014        12603                   NaN    36105           NaN   
mean      NaN          NaN           2003.459462      NaN    117.171050   
std       NaN          NaN              7.571165      NaN    200.386415   
min       NaN          NaN           1910.000000      NaN      0.000000   
25%       NaN          NaN           1999.000000      NaN     71.000000   
50%       NaN          NaN           2004.000000      NaN    107.000000   
75%       NaN          NaN           2008.000000      NaN    150.000000   
max       NaN          NaN           2019.000000      NaN  17700.000000   

        model    odometer_km  month_of_registration fuel_type       brand  \
count   46103   48559.000000           48559.000000     44538       48559   
unique    245            NaN                    NaN         7          40   
top      golf            NaN                    NaN    benzin  volkswagen   
freq     3898            NaN                    NaN     29370       10334   
mean      NaN  125815.193888               5.783212       NaN         NaN   
std       NaN   39727.711036               3.685103       NaN         NaN   
min       NaN    5000.000000               0.000000       NaN         NaN   
25%       NaN  125000.000000               3.000000       NaN         NaN   
50%       NaN  150000.000000               6.000000       NaN         NaN   
75%       NaN  150000.000000               9.000000       NaN         NaN   
max       NaN  150000.000000              12.000000       NaN         NaN   

       not_repaired_damage         date_created  nr_of_pictures   postal_code  \
count                39468                48559         48559.0  48559.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 34778                 1886             NaN           NaN   
mean                   NaN                  NaN             0.0  50975.925822   
std                    NaN                  NaN             0.0  25751.362448   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30656.000000   
50%                    NaN                  NaN             0.0  49716.000000   
75%                    NaN                  NaN             0.0  71665.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 48559  
unique                38467  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                  

### Clean year of registration
Based on description of years, those years lying outside 1910 and 2019 should be removed. The cleaned data set now  has from 48579 down to `48559` rows. 

Let us assign the new set of cleaned data to our working dataframe, `new_autos`.

In [94]:
new_autos = new2_autos

**distribution of year of registration- ** shows the highest number of vehicles belong to year 2000, amounting to 6.5%, followed by 2005 and 1999 both 6%.

In [95]:
new_autos['year_of_registration'].value_counts(normalize=True)

2000    0.064993
2005    0.060463
1999    0.059701
2004    0.055664
2003    0.055582
2006    0.055005
2001    0.054305
2002    0.051195
1998    0.048662
2007    0.046891
2008    0.045615
2009    0.042958
1997    0.040178
2011    0.033423
2010    0.032723
2017    0.028687
1996    0.028275
2012    0.026977
1995    0.025289
2016    0.025124
2013    0.016537
2014    0.013674
1994    0.012953
2018    0.009700
1993    0.008752
2015    0.008073
1992    0.007640
1990    0.007146
1991    0.006981
1989    0.003583
          ...   
1976    0.000432
1969    0.000391
1975    0.000371
1965    0.000350
1964    0.000247
1963    0.000165
1959    0.000124
1961    0.000124
1910    0.000103
1937    0.000082
1962    0.000082
1958    0.000082
1956    0.000082
1950    0.000062
1955    0.000041
2019    0.000041
1934    0.000041
1957    0.000041
1951    0.000041
1941    0.000041
1954    0.000041
1953    0.000021
1948    0.000021
1927    0.000021
1943    0.000021
1929    0.000021
1939    0.000021
1938    0.0000

### Date crawled
Looking at only the date part of timestamps, we have in this column 34 unique values. The *highest* entries belong to day 3rd april. Since this has all the 48559 values of the table, this column as no missing values in the distribution.

In [96]:
new_autos['date_crawled'].str[:10].describe()

count          48559
unique            34
top       2016-04-03
freq            1874
Name: date_crawled, dtype: object

In [97]:
new_autos['date_crawled'].str[:10].value_counts(ascending=False)

2016-04-03    1874
2016-03-20    1839
2016-03-21    1817
2016-03-12    1793
2016-03-14    1775
2016-04-04    1773
2016-03-07    1749
2016-04-02    1722
2016-03-28    1692
2016-03-19    1689
2016-03-15    1664
2016-03-29    1655
2016-03-30    1636
2016-04-01    1634
2016-03-08    1619
2016-03-09    1608
2016-03-22    1604
2016-03-11    1581
2016-03-26    1564
2016-03-23    1563
2016-03-10    1563
2016-03-31    1546
2016-03-17    1535
2016-03-25    1535
2016-03-27    1510
2016-03-16    1438
2016-03-24    1425
2016-03-05    1229
2016-03-13     761
2016-03-06     681
2016-04-05     636
2016-03-18     627
2016-04-06     154
2016-04-07      68
Name: date_crawled, dtype: int64

### Ad created


In [98]:
new_autos['date_created'].str[:10].describe()

count          48559
unique            76
top       2016-04-03
freq            1886
Name: date_created, dtype: object

In [99]:
new_autos['date_created'].str[:10].value_counts().sort_index()

2015-06-11       1
2015-08-10       1
2015-09-09       1
2015-11-10       1
2015-12-05       1
2015-12-30       1
2016-01-03       1
2016-01-07       1
2016-01-10       2
2016-01-13       1
2016-01-14       1
2016-01-16       1
2016-01-22       1
2016-01-27       3
2016-01-29       1
2016-02-01       1
2016-02-02       2
2016-02-05       2
2016-02-07       1
2016-02-08       1
2016-02-09       1
2016-02-11       1
2016-02-12       2
2016-02-14       2
2016-02-16       1
2016-02-17       1
2016-02-18       2
2016-02-19       3
2016-02-20       2
2016-02-21       3
              ... 
2016-03-09    1611
2016-03-10    1549
2016-03-11    1597
2016-03-12    1785
2016-03-13     826
2016-03-14    1709
2016-03-15    1651
2016-03-16    1463
2016-03-17    1518
2016-03-18     660
2016-03-19    1636
2016-03-20    1842
2016-03-21    1827
2016-03-22    1595
2016-03-23    1555
2016-03-24    1422
2016-03-25    1542
2016-03-26    1567
2016-03-27    1505
2016-03-28    1698
2016-03-29    1652
2016-03-30  

In [100]:
new_autos['date_created'].str[:10].value_counts(ascending=False)

2016-04-03    1886
2016-03-20    1842
2016-03-21    1827
2016-04-04    1791
2016-03-12    1785
2016-03-14    1709
2016-04-02    1706
2016-03-28    1698
2016-03-07    1687
2016-03-29    1652
2016-03-15    1651
2016-03-19    1636
2016-04-01    1634
2016-03-30    1627
2016-03-08    1620
2016-03-09    1611
2016-03-11    1597
2016-03-22    1595
2016-03-26    1567
2016-03-23    1555
2016-03-10    1549
2016-03-31    1548
2016-03-25    1542
2016-03-17    1518
2016-03-27    1505
2016-03-16    1463
2016-03-24    1422
2016-03-05    1111
2016-03-13     826
2016-03-06     743
              ... 
2016-02-05       2
2016-02-14       2
2016-02-18       2
2016-02-20       2
2016-01-10       2
2016-02-12       2
2016-02-24       2
2016-02-02       2
2016-02-26       2
2016-02-09       1
2016-02-17       1
2016-01-13       1
2016-01-14       1
2016-02-08       1
2016-01-16       1
2016-02-07       1
2016-02-11       1
2016-02-01       1
2015-11-10       1
2015-06-11       1
2016-02-22       1
2016-01-29  

**Review -** In the date created column, we have analysed above, we see that these peak around 3rd of April. These are *1,886* ads in just single day. And in the early days, there are only one or two ad created per day. 
There are also 76 unique values, total 48559, same as table. 

### Last seen
Let us look at the values in last seen column.

In [101]:
new_autos['last_seen'].str[:10].describe()

count          48559
unique            34
top       2016-04-06
freq           10771
Name: last_seen, dtype: object

In [102]:
new_autos['last_seen'].str[:10].value_counts(ascending=False)

2016-04-06    10771
2016-04-07     6405
2016-04-05     6057
2016-03-17     1363
2016-04-03     1224
2016-04-02     1211
2016-03-30     1202
2016-04-04     1188
2016-03-12     1155
2016-03-31     1154
2016-04-01     1106
2016-03-29     1086
2016-03-22     1038
2016-03-28     1014
2016-03-21     1003
2016-03-20     1003
2016-03-24      960
2016-03-25      933
2016-03-23      900
2016-03-26      816
2016-03-16      799
2016-03-15      771
2016-03-19      768
2016-03-27      760
2016-03-14      612
2016-03-11      601
2016-03-10      518
2016-03-09      467
2016-03-13      432
2016-03-08      361
2016-03-18      357
2016-03-07      262
2016-03-06      210
2016-03-05       52
Name: last_seen, dtype: int64

**Review-** We have 34 unique values, the highest being 6th April, 10771. The lowest is 5th March, only 52. Also, trend in the number of seeing/visits is that it is gradually growing from March to April, with inherent highs and lows. 

In [103]:
new_autos['last_seen'].str[:10].value_counts().sort_index()

2016-03-05       52
2016-03-06      210
2016-03-07      262
2016-03-08      361
2016-03-09      467
2016-03-10      518
2016-03-11      601
2016-03-12     1155
2016-03-13      432
2016-03-14      612
2016-03-15      771
2016-03-16      799
2016-03-17     1363
2016-03-18      357
2016-03-19      768
2016-03-20     1003
2016-03-21     1003
2016-03-22     1038
2016-03-23      900
2016-03-24      960
2016-03-25      933
2016-03-26      816
2016-03-27      760
2016-03-28     1014
2016-03-29     1086
2016-03-30     1202
2016-03-31     1154
2016-04-01     1106
2016-04-02     1211
2016-04-03     1224
2016-04-04     1188
2016-04-05     6057
2016-04-06    10771
2016-04-07     6405
Name: last_seen, dtype: int64

### Brand


In [104]:
brand_unique = new_autos['brand'].unique()
brand_unique = brand_unique[0:20]
print(brand_unique)
new_autos['brand'].value_counts(normalize=True)

['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'sonstige_autos' 'opel' 'mazda' 'porsche' 'mini'
 'toyota' 'dacia' 'nissan' 'jeep' 'saab']


volkswagen        0.212813
bmw               0.108631
opel              0.108631
mercedes_benz     0.095780
audi              0.085834
ford              0.069688
renault           0.047880
peugeot           0.029449
fiat              0.026010
seat              0.018925
skoda             0.016063
nissan            0.015260
mazda             0.015219
smart             0.014292
citroen           0.014127
toyota            0.012583
hyundai           0.009947
sonstige_autos    0.009700
volvo             0.009061
mini              0.008608
mitsubishi        0.008176
honda             0.007990
kia               0.007105
alfa_romeo        0.006611
porsche           0.005910
suzuki            0.005890
chevrolet         0.005663
chrysler          0.003480
dacia             0.002657
daihatsu          0.002512
jeep              0.002204
subaru            0.002101
land_rover        0.002039
saab              0.001627
daewoo            0.001565
jaguar            0.001524
trabant           0.001380
r

In [105]:
# descending list of popular brands
brand_selected = new_autos['brand'].value_counts(ascending=False).index

In [106]:
brand_selected

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini',
       'mitsubishi', 'honda', 'kia', 'alfa_romeo', 'porsche', 'suzuki',
       'chevrolet', 'chrysler', 'dacia', 'daihatsu', 'jeep', 'subaru',
       'land_rover', 'saab', 'daewoo', 'jaguar', 'trabant', 'rover', 'lancia',
       'lada'],
      dtype='object')

In [107]:
brand_topten = brand_selected[0:9]

## Top ten Brands

From analysis of cleaned data set we can say about the share of market , the ten **most popular** brands. 

The **highest** being *Volkswagen* at **`20%`**. 

This is followed by *BMW, opel, and mercedez benz*, nearly `10%` each.

In [108]:
brand_topten

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat'],
      dtype='object')

### Average price of topten

Having known the major brands, let us see their average prices of top ten brands. 

**Volkwagen is around , in dollar 6000, BMW  8500, opel  5000.**

    - audi   9212.9306621881,  
    - mercedes_benz   30025.542248978712,  
    - renault   2431.195698924731,  
    - peugeot   3065.611888111888,  
    - bmw   8493.857251184834,  
    - volkswagen   6600.382233404296,  
    - ford   7305.141843971631,  
    - opel   5281.321706161138,  
    - fiat   12566.541567695962

In [110]:
brand_avgprice = {}
brand_avgmil = {}

for b in brand_topten:
    select_rows = new_autos[new_autos['brand'] == b]
    average_price = select_rows['price_$'].mean()
    brand_avgprice[b] = average_price
    average_mil = select_rows['odometer_km'].mean()
    brand_avgmil[b] = average_mil
    
print(brand_avgprice)
print(brand_avgmil)

    

{'audi': 9212.9306621881, 'mercedes_benz': 30025.542248978712, 'renault': 2431.195698924731, 'peugeot': 3065.611888111888, 'bmw': 8493.857251184834, 'volkswagen': 6600.382233404296, 'ford': 7305.141843971631, 'opel': 5281.321706161138, 'fiat': 12566.541567695962}
{'audi': 129492.56238003839, 'mercedes_benz': 130852.50483766932, 'renault': 128062.36559139784, 'peugeot': 127356.64335664336, 'bmw': 132686.2559241706, 'volkswagen': 128961.67989161989, 'ford': 124327.71867612294, 'opel': 129455.92417061611, 'fiat': 117593.03246239113}


In [112]:
bmp_series = pd.Series(brand_avgprice)
bmm_series = pd.Series(brand_avgmil)
print(bmp_series)
print(bmm_series)

audi              9212.930662
bmw               8493.857251
fiat             12566.541568
ford              7305.141844
mercedes_benz    30025.542249
opel              5281.321706
peugeot           3065.611888
renault           2431.195699
volkswagen        6600.382233
dtype: float64
audi             129492.562380
bmw              132686.255924
fiat             117593.032462
ford             124327.718676
mercedes_benz    130852.504838
opel             129455.924171
peugeot          127356.643357
renault          128062.365591
volkswagen       128961.679892
dtype: float64


In [113]:
top_ten_table = pd.DataFrame(bmp_series, columns=['mean_price'])

***Conclusion***

## Of top ten brands
### Mean price and Mean Mileage
Now that we have the analysis - the calculated values **about price** with calculated values **about mileage**, we just need to put them together with brand.

This has been presented in table below. 



In [117]:
top_ten_table['mean_mileage'] = bmm_series
print(top_ten_table)

                 mean_price   mean_mileage
audi            9212.930662  129492.562380
bmw             8493.857251  132686.255924
fiat           12566.541568  117593.032462
ford            7305.141844  124327.718676
mercedes_benz  30025.542249  130852.504838
opel            5281.321706  129455.924171
peugeot         3065.611888  127356.643357
renault         2431.195699  128062.365591
volkswagen      6600.382233  128961.679892


**We can see `mean mileage` across all the brands is nearly the same, around 125,000km and 10,000km above or below. However, the `price` has much more variation. Renault selling for lowest average price, and Mercedez_benz at highest average price.**